### ISSUE TITLE
Async operations are clogging the loop?

### ISSUE THREAD

**Issue by jeromegn** [laverdet/isolated-vm#25]
I've been using isolated-vm to process a stream with custom JS functions which I run within a V8 instance, but I'm encountering sluggishness in places.

It goes something like:
- get a Readable stream
- create a Transform stream which passes the input to a function running within the isolated vm
- wait for callback to be called with desired output
- push into the stream
- call the Transform callback

```javascript
const tr = new Transform({
  transform(chunk, encoding, callback) {
    someIsolatedFunction.apply(null, [chunk.toString(), new ivm.Reference(function(err, newChunk){
      // error checking
      tr.push(newChunk)
      callback()
    }]
})

src.pipe(tr).pipe(dst)
```

There is major slowness if I don't use a `setTimeout(fn, 0)` or `setImmediate(fn)` for the `callback()`.

This is only an example, I've noticed that in many places.

I run everything using the async functions provided by isolated-vm. Everything works great except for slowness that can reach up to 1-5 seconds (during which nothing is happening.)

I'm sure I'm missing something or doing something incorrectly. If this is normal and I should pepper my code with `setImmediate`, that's ok with me, but I'm not sure where to put them.

**Comment by laverdet** [comment]
Is there any chance you could distill this down into an encapsulated test case I could easily run on my machine?

**Comment by jeromegn** [comment]
There is a chance :smile:

I tried the other than and couldn't replicate it in isolation. Meaning something, somewhere, was causing the blockage.

I guess my question is more generic: Is it possible that asynchronously calling an `ivm.Reference<Function>` from an isolated v8 could block something?

I've noticed the slowness happens when my isolated v8 code will apply a function, even though I'm not doing `await fn.apply()`, it'll still block the loop somehow unless I immediately return from the called function.

Another example of this behaviour is a `fetch` function I'm exposing to the isolated vm. It accepts a callback (because I can't return a native node.js Promise afaict). If I don't use `setImmediate` right await, it'll somehow block for a long time. Even after I've called the callback, it'll take more time for the isolated vm's tasks to be picked up by the event loop (I think.)

However, I can't say I have been able to replicate those in isolation. Which is very weird. I'm not using any of the synchronous APIs from node.js in my code. I either use callbacks or promises.

Any general pointers to help fix what might cause this would help. I know this is not much to go on.

I'd be happy to show you some code in private. No need for a NDA or anything, I just don't want to expose that code publicly just yet!

**Comment by laverdet** [comment]
None of the async functions should ever block for too long. There's two things I can think of that could be happening here:

**Situation 1**
There is some kind of lock contention happening which prevents the async operation from being queued up correctly. All the async functions will eventually make it to the C++ function `ShareableIsolate->ScheduleTask()` in `shareable_isolate.h`. It's a simple function that just adds a function pointer to the end of a queue. If this function blocking for some reason it would be the line with the `std::unique_lock`.

**Situation 2**
Somehow the event loop is not actually blocked but is just sleeping. You can test this by adding `setInterval(() => console.log(new Date), 500)` in your application somewhere. If the console.log calls don't stop then you know that the main event loop isn't actually stopped.

**Comment by jeromegn** [comment]
Ok, I've conducted more tests and it appears the slowness in this specific streaming case was because the receiver is slow to accept writes. I'm not sure, but adding that `setImmediate` makes a massive difference (10x faster with it.) It doesn't make much sense to me since `callback()` seems to return immediately anyway.

I did notice slowdowns before, however I can't replicate them anymore. I will reopen this issue if I can replicate in an isolated test.
